### PR TITLE
Grouping on arrays as arrays

### PR DESCRIPTION
**Pull request by cryptoe**
### Description
Currently, grouping on multivalue columns inside druid work by exploding each value. From the druid [docs](https://druid.apache.org/docs/latest/querying/multi-value-dimensions.html#grouping) : 

For an example datasoure 
```json
{"timestamp": "2011-01-12T00:00:00.000Z", "tags": ["t1","t2","t3"]}  #row1
{"timestamp": "2011-01-13T00:00:00.000Z", "tags": ["t3","t4","t5"]}  #row2
{"timestamp": "2011-01-14T00:00:00.000Z", "tags": ["t5","t6","t7"]}  #row3
{"timestamp": "2011-01-14T00:00:00.000Z", "tags": []}                #row4
```

```json
{
  "queryType": "groupBy",
  "dataSource": "test",
  "intervals": [
    "1970-01-01T00:00:00.000Z/3000-01-01T00:00:00.000Z"
  ],
  "granularity": {
    "type": "all"
  },
  "dimensions": [
    {
      "type": "default",
      "dimension": "tags",
      "outputName": "tags"
    }
  ],
  "aggregations": [
    {
      "type": "count",
      "name": "count"
    }
  ]
}
```

This query returns the following result:

```json
[
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "t1"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "t2"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 2,
      "tags": "t3"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "t4"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 2,
      "tags": "t5"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "t6"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "t7"
    }
  }
]
```

Notice that original rows are "**exploded**" into multiple rows and merged.

The goal of this PR is to group on multi value columns as arrays without exploding them. Please look at the virtual column and the dimensionSpec to see how we are activating this behavior. For eg:

```json
{
  "queryType": "groupBy",
  "dataSource": "test",
  "intervals": [
    "1970-01-01T00:00:00.000Z/3000-01-01T00:00:00.000Z"
  ],
  "granularity": {
    "type": "all"
  },
  "virtualColumns" : [ {
    "type" : "expression",
    "name" : "v0",
    "expression" : "mv_to_array(\"tags\")",
    "outputType" : "ARRAY<STRING>"
  } ],
  "dimensions": [
    {
      "type": "default",
      "dimension": "v0",
      "outputName": "tags"
      "outputType":"ARRAY<STRING>"
    }
  ],
  "aggregations": [
    {
      "type": "count",
      "name": "count"
    }
  ]
}
```
will return with the following results 

```
[
 {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "[]"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "["t1","t2","t3"]"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 1,
      "tags": "[t3","t4","t5"]"
    }
  },
  {
    "timestamp": "1970-01-01T00:00:00.000Z",
    "event": {
      "count": 2,
      "tags": "["t5","t6","t7"]"
    }
  }
]
```

To activate this behavior in SQL we use a new `mv_to_array` which takes in a multiValue/String column. 

```sql
select mv_to_array[tags] , count(*) from inline_data group by 1
```

### Core Engine changes
In druid, we were coercing array's to string in the calcites layer. With this PR, we have removed the coercion and arrays are being passed as arrays to the native layer.

The idea was to copy concepts from how `DictionaryBuildingStringGroupByColumnSelectorStrategy` generated int<-> List<T> on the fly. 
Some optimizations are done for Array[String] while generating the dictionary keeping in mind that we want to store each string only once on the heap and reference that with the integer. As we are dealing with arrays, we are referencing the indexed array with yet another int so that the least amount of heap is used when string cardinality is low. 
So if we have 2 rows:
```
 [a,b]
 [b,c]
```
the global dictionary would look like this:
```
"a"<>1
"b"<>2
"c"<>3
```
and the corresponding arrays would look like
``` 
[1,2]<>1
[2,3]<>2 
```

Also the inMemory structures needed to implement `comparable` and would be called per row. Hence `ComparableStringArrays`, `ComparableIntArrays` are coded in such a way that they don't have the overhead of `Lists` and `Integers`. 

<hr>

##### Key changed/added classes in this PR
 * `ComparableStringArray.java`
 * `ComparableIntArray.java`
 * `ResultRowDeserializer.java`
 * `RowBasedGrouperHelper.java`
 * `ArrayConstructorOperatorConversion.java`
 * `ArrayStringGroupByColumnSelectorStrategy.java`
 * `ListGroupByColumnSelectorStrategy.java`
 * `ComparableList.java`
 * `Calcites.java`
 * `Function.java`

<hr>

To enable the legacy behavior, for whatever reason, one can set a runtime property  
`-Ddruid.expressions.processArraysAsMultiValueStrings=true` while starting the broker. 
 

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by FrankChen021** [comment]
I don't think putting the output format into the data source spec is a good design. Is it possible to extend current `ResultFormat` to meet your need?

**Comment by cryptoe** [comment]
> 

Hey @FrankChen021 Thanks for looking into it. I am not sure that I understand your concern. We generally put the column type in the dimension spec no. Checkout the class `DefaultDimensionSpec`. I am using that.

The pr extends the functionality of the group by engine. The grouping keys themselves are changing. I am not sure how `resultFormat` would help here.

**Comment by dbardbar** [comment]
Just wondering - can't you achieve the same thing with MV_TO_STRING? 
Also, have you given any thought about how your code will be generated from SQL?

**Comment by cryptoe** [comment]
@dbardbar Good Question. The goal here is to make druid SQL layer similar to other SQL layers . In traditional DB's like POSTGRES , the SQL construct of declaring an array is an 'array'. I am using that construct itself to hook the SQL layer with the native layer. Just finishing up some test cases for that.

**Comment by dbardbar** [comment]
@cryptoe - two questions
1) Would this work also for a VirtualColumn which produces a multi-value? Specifically, VirtualColumn of type "mv-filtered".
2) We needed the capability in this PR, but since it was missing we created another string dimension during ingestion, which holds the values of the MV column as a simple string. We then used that extra column in our GROUP BY, to achieve the same functionality as this PR. Do you think that your method would be comparable in performance?

**Comment by cryptoe** [comment]
>     1. Would this work also for a VirtualColumn which produces a multi-value? Specifically, VirtualColumn of type "mv-filtered".
 Is it possible to share a sample Q here. I am trying to understand the use case here. Do you want to group by on array<array> ?
 https://github.com/apache/druid/pull/12078/files#diff-8bc53aec44924e671b3c6f6f34c6f0c499f873d9000649c47c237444707aea4bR975 . This PR does support virtual col's .


>     2. We needed the capability in this PR, but since it was missing we created another string dimension during ingestion, which holds the values of the MV column as a simple string. We then used that extra column in our GROUP BY, to achieve the same functionality as this PR. Do you think that your method would be comparable in performance?

If the original string dimension is of low cardinality, then IMHO this implementation will be slightly more performant with regards to memory usage across the historical's, brokers as we are doing optimizations on the way we lookup.

**Comment by dbardbar** [comment]
@cryptoe,

The use-case is that some of the values in the MV field are not interesting for a specific query, and we would like to ignore them for the purpose of the GROUP BY. They are kept there because those ignored tags might be used for filtering, or might be used for GROUP BY when performing a different query.

An example query, using the example data appearing at the top of this PR:

SELECT MV_FILTER_ONLY(tags, ARRAY['t3', 't4'], COUNT(*) FROM test GROUP BY 1

with your new code enabled, I would expect the following to return:
```
["t3"],       1            (from row1)
["t3", "t4"], 1            (from row2)
[],           2            (from row3+row4)
```

**Review by clintropolis — COMMENTED** [review]
thanks for looking into this, super into being able to group on arrays natively :+1:

**Comment by FrankChen021** [comment]
@cryptoe It's my fault to misunderstand your design. Just ignore the comment I left before.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java [hunk 1]
```diff
@@ -1357,6 +1373,12 @@ private RowBasedKeySerdeHelper makeSerdeHelper(
     )
     {
       switch (valueType.getType()) {
+        case ARRAY:
+          return new ArrayRowBasedKeySerdeHelper(
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:1377]: Acked. Working on it.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java [hunk 2]
```diff
@@ -386,13 +393,16 @@ public static boolean canPushDownLimit(ColumnSelectorFactory columnSelectorFacto
     @Override
     public GroupByColumnSelectorStrategy makeColumnSelectorStrategy(
         ColumnCapabilities capabilities,
-        ColumnValueSelector selector
+        ColumnValueSelector selector,
+        DimensionSpec dimensionSpec
     )
     {
       switch (capabilities.getType()) {
         case STRING:
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:401]: I will be breaking up the PR into two phases.
1. Add the string implementation first but also have the generic building blocks in place.
2. Add the rest of the world implementation. 
At this point, I am not sure. If one PR or 2 PR's would be necessary.

File: processing/src/main/java/org/apache/druid/query/groupby/ResultRowDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+              final List<String> strings = new ArrayList<>();
+              while (jp.nextToken() != JsonToken.END_ARRAY) {
+                strings.add(jp.getText());
+              }
+              retVal[i] = ComparableStringArray.of(strings.toArray(new String[0]));
+              break;
+            }
+
+          default:
+            throw new ISE("Can't handle type [%s]", types.get(i).asTypeString());
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/query/groupby/ResultRowDeserializer.java:113]: Fair enough. I can remove this class altogether. Added it initially to have stricter types while deserializing. But if we have unknown types then falling back on Jackson seems much better.

**Review by cryptoe — COMMENTED** [review]
Thanks for the Review

File: processing/src/main/java/org/apache/druid/query/dimension/ColumnSelectorStrategyFactory.java [hunk 4]
```diff
@@ -24,5 +24,9 @@
 
 public interface ColumnSelectorStrategyFactory<ColumnSelectorStrategyClass extends ColumnSelectorStrategy>
 {
-  ColumnSelectorStrategyClass makeColumnSelectorStrategy(ColumnCapabilities capabilities, ColumnValueSelector selector);
+  ColumnSelectorStrategyClass makeColumnSelectorStrategy(
+      ColumnCapabilities capabilities,
+      ColumnValueSelector selector,
+      DimensionSpec dimensionSpec
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/query/dimension/ColumnSelectorStrategyFactory.java:30]: Fixed!

**Review by cryptoe — COMMENTED** [review]
Moving to PR out of draft

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 5]
```diff
@@ -1306,6 +1308,321 @@ public void testMultiValueDimension()
...
+            245L
+        )
+    );
+
+    Iterable<ResultRow> results = GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
+    TestHelper.assertExpectedObjects(expectedResults, results, "multi-value-dims-groupby-arrays");
+  }
+
+  @Test
+  public void testMultiValueDimensionAsArrayWithVirtualDim()
```
Comment (gianm) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1488]: Seems like the same test as testMultiValueDimensionAsArrayWithOtherDims.

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 6]
```diff
@@ -1306,6 +1308,321 @@ public void testMultiValueDimension()
     TestHelper.assertExpectedObjects(expectedResults, results, "multi-value-dim");
   }
 
+  @Test
+  public void testMultiValueDimensionAsArray()
```
Comment (gianm) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1312]: Please include other tests for:

- Dimension type STRING_ARRAY on an underlying column of type STRING that happens to be single-valued, like `placement`
- Dimension type STRING_ARRAY on an underlying column of type STRING that happens to be multi-valued, like `placementish`
- Dimension type STRING_ARRAY on an underlying column of numeric type, like `index`
- Dimension type STRING on an underlying column of type STRING_ARRAY (such as a virtual column)
- An extraction dimension spec of type STRING on a column of type STRING_ARRAY, like `new ExtractionDimensionSpec("v0", "alias", ColumnType.STRING, new SubstringDimExtractionFn(1, 1))`
- An extraction dimension spec of type STRING_ARRAY on a column of type STRING_ARRAY, like `new ExtractionDimensionSpec("v0", "alias", ColumnType.STRING_ARRAY, new SubstringDimExtractionFn(1, 1))`
- A virtual column that builds an array from a numeric type, like `array(index)`, and declares its output type as STRING_ARRAY
- A nested groupBy query, where the inner and outer queries both have dimensions of type STRING_ARRAY

If any of these aren't supposed to work, use `expectedException` to verify that the exception message is good. Otherwise, verify that the results are what you expect to see.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java [hunk 7]
```diff
@@ -128,27 +128,10 @@ public static String escapeStringLiteral(final String s)
...
   /**
-   * Convert {@link RelDataType} to the most appropriate {@link ValueType}, coercing all ARRAY types to STRING (until
-   * the time is right and we are more comfortable handling Druid ARRAY types in all parts of the engine).
-   *
-   * Callers who are not scared of ARRAY types should isntead call {@link #getValueTypeForRelDataTypeFull(RelDataType)},
-   * which returns the most accurate conversion of {@link RelDataType} to {@link ValueType}.
+   * Convert {@link RelDataType} to the most appropriate {@link ValueType}}.
    */
   @Nullable
   public static ColumnType getColumnTypeForRelDataType(final RelDataType type)
```
Comment (gianm) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:134]: I take it that no callers are "scared of ARRAY types" anymore.

1. What are the main risks of making this change, & are they mitigated by testing?
2. Do we need a feature flag?

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 8]
```diff
@@ -403,7 +402,9 @@ private static Grouping computeGrouping(
       }
 
       final RelDataType dataType = rexNode.getType();
-      final ColumnType outputType = Calcites.getColumnTypeForRelDataType(dataType);
+      final ColumnType outputType;
+      outputType = Calcites.getColumnTypeForRelDataType(dataType);
```
Comment (gianm) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:406]: Needless change?

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java [hunk 9]
```diff
@@ -143,54 +144,6 @@ public void testSelectNonConstantArrayExpressionFromTable() throws Exception
     );
   }
 
-  @Test
-  public void testSelectNonConstantArrayExpressionFromTableForMultival() throws Exception
```
Comment (gianm) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:147]: Why did you delete this test?

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java [hunk 10]
```diff
@@ -923,6 +876,277 @@ public void testArrayOffset() throws Exception
...
+        ),
+        ImmutableList.of(
+            new Object[]{ImmutableList.of(0.0), 4L},
+            new Object[]{ImmutableList.of(0.10000000149011612), 1L},
+            new Object[]{ImmutableList.of(1.0), 1L}
+        )
+    );
+  }
+
+  @Test
```
Comment (gianm) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:1061]: Please include tests for the `ARRAY` constructor with multiple arguments, like:

- two multi-valued columns: `ARRAY[dim3, dim3]`
- multi-valued plus regular column: `ARRAY[dim3, dim1]`

**Review by gianm — COMMENTED** [review]
FYI: I reviewed the overall design & testing approach. I didn't review the implementation on the native side very closely.

I think it largely looks good, but since it's a major new piece of functionality, IMO it needs more testing than it has. In the comments I wrote some recommendations for additional tests.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java (hunk 7 above)
Comment (clintropolis) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:134]: the engine has been waiting on array grouping for this coercion to go away, though there is a missing piece I think where we need to ensure that when grouping on a virtual column we never plan to a topn query in `DruidQuery`.

I think most of the `MV_` prefixed functions have their return type as `VARCHAR` rather than `ARRAY`, so in theory the "documented" multi-value string functions shouldn't actually be affected, and any that will be is a bug that we should fix. The undocumented `ARRAY_` prefixed functions will now behave as intended and validate/operate as true SQL arrays (and so group correctly as arrays). If anyone was using the undocumented `ARRAY_` prefix functions, then I guess that behavior will change, since grouping on those expressions would previously have been grouping as `STRING` due to the implicit unnest.

The only reason I can think of to add a feature flag is if need a way to hide bugs that this change causes, and that _should_ only happen if any of the `MV_` functions have bugs in their SQL output types, or if people were using the undocumented `ARRAY_` functions.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java (hunk 9 above)
Comment (clintropolis) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:147]: heh, i wrote this test originally to document moderately strange current behavior. This PR changes the behavior of `array` constructor function to no longer "map" multi-value string inputs and instead treat them as arrays, so it should probably be repurposed to test whatever the new behavior should be probably

The map behavior was not very intuitive, since it would make a nested array where the elements are single element arrays from the multi-value string, so idk that it needs preserved.

We might want to consider making use of a multi-value string inside of the array constructor an error, since i'm not sure there is an intuitive behavior because it is ambiguous whether the user is acknowledging the column as a multi-value, or doesn't know and thinks it single valued, which most of the other functions don't have this ambiguity, though it probably doesn't need to be changed in this PR.

File: core/src/main/java/org/apache/druid/math/expr/Function.java [hunk 11]
```diff
@@ -2994,8 +2994,14 @@ public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
       Object[] out = new Object[length];
 
       ExpressionType arrayType = null;
+
       for (int i = 0; i < length; i++) {
         ExprEval<?> evaluated = args.get(i).eval(bindings);
+        // short circuit the case where you have one input argument whose input type is array. So array function has
```
Comment (clintropolis) [inline core/src/main/java/org/apache/druid/math/expr/Function.java:3000]: i have a concern with this change, best illustrated by the implication that now an expression like this

```
array(array('x','y','z'))
```
will now produce `['x','y','z']` instead of `[['x','y','z']]` as one would expect, or would happen if the expression were this instead:

```
array(array('x','y','z'),array('a','b','c'))
```
which will still produce `[['x','y','z'],['a','b','c']]`

Instead of hijacking this function to accommodate multi-value string behavior, I think we should instead add a `MV_TO_ARRAY` function in SQL that only accepts a string, and then in the native expression layer either just use `cast(x, 'ARRAY<STRING>')` or also add a `mv_to_array` function that basically does the same thing as cast. This way we are explicit on when we intend to treat a multi-value column directly as an array

Comment (cryptoe) [reply core/src/main/java/org/apache/druid/math/expr/Function.java:3000]: The original intention was to keep it close to SQL but i realized SQL does not have multi value col's :). So a new function may be necessary . 
Cool I am adding a new `MV_TO_ARRAY` function to both the native and the SQL layer. 
Thanks for the tip

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java [hunk 12]
```diff
@@ -403,6 +412,20 @@ public GroupByColumnSelectorStrategy makeColumnSelectorStrategy(
           return makeNullableNumericStrategy(new FloatGroupByColumnSelectorStrategy());
         case DOUBLE:
           return makeNullableNumericStrategy(new DoubleGroupByColumnSelectorStrategy());
+        case ARRAY:
+          switch (capabilities.getElementType().getType()) {
+            case LONG:
+              return new ListLongGroupByColumnSelectorStrategy();
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:418]: The reason was that ListLong internally working on a list as the DS and hence the name but seems weird here. 
Changed everything to ArrayX as I think that's a better name signifying whats the selector strategy for.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java (hunk 7 above)
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:134]: As part of this patch, then, let's check all the MV functions to make sure the return types are what we expect them to be.

As to the array stuff, one of them is documented: ARRAY_AGG. And you could do that in a subquery and group on the result. Some questions about that:

- Do we have tests for the case of grouping on the result of an ARRAY_AGG? If not we should add it.
- Does the behavior change or does this introduce excess risk due to potential bugs? If so, a feature flag would be a good idea.

File: core/src/main/java/org/apache/druid/math/expr/Function.java (hunk 11 above)
Comment (gianm) [reply core/src/main/java/org/apache/druid/math/expr/Function.java]: Makes sense to me too. MV_TO_ARRAY would be simpler than ARRAY and have fewer cases to worry about. We can start there.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java (hunk 7 above)
Comment (cryptoe) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:134]: > As part of this patch, then, let's check all the MV functions to make sure the return types are what we expect them to be.

Looks like all mv functions are overriding the output signature of the corresponding array functions to return a VARCHAR instead of an array

> 
> As to the array stuff, one of them is documented: ARRAY_AGG. And you could do that in a subquery and group on the result. Some questions about that:
> 
>     * Do we have tests for the case of grouping on the result of an ARRAY_AGG? If not we should add it.
Were you looking for [this](https://github.com/apache/druid/pull/12078/files#diff-8bc53aec44924e671b3c6f6f34c6f0c499f873d9000649c47c237444707aea4bL1640) test case ?

> 
>     * Does the behavior change or does this introduce excess risk due to potential bugs? If so, a feature flag would be a good idea.

Still thinking about this

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java [hunk 13]
```diff
@@ -340,7 +348,8 @@ public static boolean isAllSingleValueDims(
 
               // Now check column capabilities, which must be present and explicitly not multi-valued
               final ColumnCapabilities columnCapabilities = inspector.getColumnCapabilities(dimension.getDimension());
-              return columnCapabilities != null && columnCapabilities.hasMultipleValues().isFalse();
+              return dimension.getOutputType().equals(ColumnType.STRING_ARRAY)
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:351]: Done

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 14]
```diff
@@ -1306,6 +1308,321 @@ public void testMultiValueDimension()
...
+  {
+    if (config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V1)) {
+      return;
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1315]: Added expected exception

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 5 above)
Comment (cryptoe) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1492]: Good catch. Looks like duplicate test case got added during refactoring

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 6 above)
Comment (cryptoe) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1313]: Added the above mentioned test cases

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 15]
```diff
@@ -404,6 +403,7 @@ private static Grouping computeGrouping(
 
       final RelDataType dataType = rexNode.getType();
       final ColumnType outputType = Calcites.getColumnTypeForRelDataType(dataType);
+
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:406]: nit: unnecessary changes

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java [hunk 16]
```diff
@@ -432,7 +432,6 @@ public void serialize(
         }
       }
     };
-
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java:435]: nit: unecessary change

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 17]
```diff
@@ -1306,6 +1309,531 @@ public void testMultiValueDimension()
...
+
+  @Test
+  public void testNestedGroupByWithArrays()
+  {
+    if (config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V1)) {
+      expectedException.expect(UnsupportedOperationException.class);
+      expectedException.expectMessage("GroupBy v1 only supports dimensions with an outputType of STRING");
+    } else {
+      expectedException.expect(IAE.class);
+      expectedException.expectMessage("Cannot create query type helper from invalid type [ARRAY<STRING>]");
```
Comment (clintropolis) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1804]: why doesn't this query work?

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 18]
```diff
@@ -1306,6 +1309,531 @@ public void testMultiValueDimension()
...
+
+  @Test
+  public void testNumericDimAsStringArrayWithoutExpression()
+  {
+    if (config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V1)) {
+      expectedException.expect(UnsupportedOperationException.class);
+      expectedException.expectMessage("GroupBy v1 only supports dimensions with an outputType of STRING");
+    } else if (!vectorize) {
+      expectedException.expect(RuntimeException.class);
+      expectedException.expectMessage("java.lang.Double cannot be cast to [Ljava.util.Objects");
```
Comment (clintropolis) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:1617]: i wonder if we should have tighter validation on matching dimension spec with column type, if we intend to treat dimension spec output type as a sort of cast, instead of letting it hit actual class cast exceptions

File: sql/src/test/java/org/apache/druid/sql/calcite/BaseCalciteQueryTest.java [hunk 19]
```diff
@@ -138,6 +138,12 @@ public static void setupNullValues()
     NULL_STRING = NullHandling.defaultStringValue();
     NULL_FLOAT = NullHandling.defaultFloatValue();
     NULL_LONG = NullHandling.defaultLongValue();
+    NULL_LIST = new ArrayList<Object>()
+    {
+      {
+        add(null);
+      }
+    };
```
Comment (clintropolis) [inline sql/src/test/java/org/apache/druid/sql/calcite/BaseCalciteQueryTest.java:146]: hmm, I know this is because of the coercion behavior in `ExprEval`, but I'm having some regret about the decision to homogenize `null`, `[]`, and `[null]` to `[null]`. I think it made the most sense when mapping back to `STRING` type, but I think the coercion logic should probably be different when we are grouping on a multi-value string as an ARRAY<STRING> instead of a STRING, and the input bindings should homogenize the value to `null` instead in that case.

We don't need to change this right now I think, since this PR still isn't documenting this functionality, but something to think about for the future I think, since `[null]` is sort of strange on the result side of things when stuff is left as an actual array

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java (hunk 7 above)
Comment (cryptoe) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:134]: Adding a flag in the query Context and take a decision based on that may not be trivial as the code will need to be extensively refactored.
For eg : checkout `RowSignature#fromRelDataType` https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/table/RowSignatures.java
I am not sure if adding changes to a few places where the QueryContext is passable and where its not passable, convert to array without honoring the flag, has on the query Engine. Especially in the case where the user has set the flag and half of the places we are honoring the flag and the other half we are not.

Therefore adding a RuntimeProperty in `ExpressionProcessingConfig`

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 17 above)
Comment (cryptoe) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java]: fixed

**Comment by cryptoe** [comment]
Ordering on array grouping keys is not fully supported. I was hoping to finish that in a subsequent PR.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java (hunk 7 above)
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:137]: > Were you looking for [this](https://github.com/apache/druid/pull/12078/files#diff-8bc53aec44924e671b3c6f6f34c6f0c499f873d9000649c47c237444707aea4bL1640) test case ?

Yes! thanks!

**Comment by gianm** [comment]
@cryptoe could you please post a current list of what you intend to do as near-term follow-up PRs? I'm asking since while reading through the latest state of the patch, I had a hard time keeping track of what was "done", what was "not done", and what missing stuff is "soon to come".

**Comment by cryptoe** [comment]
Sure. Collating all the next steps:
- Getting ordering on array grouping key to work
- Null coercion
- mv_to_array to support expressions 
- Query context flag to not allow grouping on multi value columns
- Refactor stuff to remove DimensionHandlerUtils returning a comparable
- Tighter validation on matching dimension spec with column type

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/ArraySqlAggregator.java [hunk 20]
```diff
@@ -114,7 +113,7 @@ public Aggregation toDruidAggregation(
 
     final String fieldName;
     final String initialvalue;
-    final ColumnType druidType = Calcites.getValueTypeForRelDataTypeFull(aggregateCall.getType());
+    final ColumnType druidType = Calcites.getColumnTypeForRelDataType(aggregateCall.getType());
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/ArraySqlAggregator.java:116]: we should maybe still call the old version of this method so that behavior doesn't change when the flag is set to the old behavior (which would change the behavior of this method)

Comment (cryptoe) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/ArraySqlAggregator.java:116]: good catch. Fixed.

**Review by clintropolis — APPROVED** [review]
lot of stuff I can think to do as follow-ups, but the basic behavior lgtm, i think this is a good start :+1:
